aten/src/ATen/native/quantized/cpu/qfc_prepack.cpp

```diff
@@ -0,0 +1,108 @@
+#include <ATen/ATen.h>
+#include <ATen/core/op_registration/op_registration.h>
+// #include <ATen/quantized/Quantizer.h>
+
+#include <algorithm>
+#include <tuple>
+
+#include "fbgemm/Fbgemm.h"
+#include "fbgemm/QuantUtils.h"
+
+namespace at {
+namespace native {
+namespace {
+
+class QFCPackWeightInt8 final : public c10::OperatorKernel {
+ public:
+  // Calculate the column offsets
+  // Note this includes the sum of the columns as well as the scalar term
+  // B_zero_point * K, whereas the row_offsets created by
+  // PackAWithQuantRowOffset is only the sum of the A rows.
+  void calc_col_offsets_transpose(
+      int K,
+      int N,
+      const int8_t* Bint8,
+      int32_t B_zero_point,
+      int32_t* col_offsets) {
+    for (size_t i = 0; i < N; ++i) {
+      int32_t sum = 0;
+      for (size_t j = 0; j < K; ++j) {
+        sum += Bint8[i * K + j];
+      }
+      col_offsets[i] = sum - B_zero_point * K;
+    }
+  }
+
+  std::tuple<at::Tensor, at::Tensor> operator()(
+      const at::Tensor& weight,
+      int64_t weight_zero_point
+      ) {
+    auto N = weight.size(0);
+    auto K = weight.size(1);
+
+    int32_t weight_zero_point_int32 = static_cast<int32_t>(weight_zero_point);
+
+    auto col_offsets =
+        at::zeros_like(weight).sum({1}).to(at::kInt).contiguous();
+    calc_col_offsets_transpose(
+        /*K=*/K,
+        /*N=*/N,
+        /*Bint8=*/weight.data<int8_t>(),
+        /*B_zero_point=*/weight_zero_point_int32,
+        /*col_offsets=*/col_offsets.data<int32_t>());
+
+
+    auto weight_contig = weight.contiguous();
+    auto contiguous_ptr = weight_contig.data<int8_t>();
+
+
+    auto* ptr = new fbgemm::PackBMatrix<int8_t>(
+        /*trans=*/fbgemm::matrix_op_t::Transpose,
+        /*nRow=*/K,
+        /*nCol=*/N,
+        /*smat=*/contiguous_ptr,
+        /*ld=*/K,
+        /*pmat=*/nullptr, // PackBMatrix manages ownership of pmat
+        /*groups=*/1);
+
+    // ptr->printPackedMatrix("packedB inside qfc_prepack: ");
+
+    // We store this instance away in a Tensor and register a deleter function
```

Comment:
Please add a TODO stating that we will need to replace this with torchscript classes at a later point.